test_result.go

```diff
@@ -0,0 +1,405 @@
+package main
+
+import (
+	"encoding/xml"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/gin-gonic/gin"
+	"github.com/iver-wharf/wharf-api/internal/ctxparser"
+	"github.com/iver-wharf/wharf-core/pkg/ginutil"
+	"github.com/iver-wharf/wharf-core/pkg/problem"
+	"gorm.io/gorm"
+)
+
+type buildTestResultModule struct {
+	Database *gorm.DB
+}
+
+type artifactListItem struct {
+	FileName   string `json:"fileName"`
+	ArtifactID uint   `json:"artifactId"`
+}
+
+// TestResultListSummary contains data about several test result files.
+type TestResultListSummary struct {
+	BuildID uint `json:"buildId"`
+	Total   uint `json:"total"`
+	Failed  uint `json:"failed"`
+	Passed  uint `json:"passed"`
+	Skipped uint `json:"skipped"`
+}
+
+func (m buildTestResultModule) Register(r gin.IRouter) {
+	testResult := r.Group("/test-result")
+	{
+		testResult.POST("/", m.postBuildTestResultDataHandler)
+
+		testResult.GET("/detail", m.getBuildAllTestResultDetailsHandler)
+
+		testResult.GET("/summary", m.getBuildAllTestResultSummariesHandler)
+		testResult.GET("/summary/:artifactId", m.getBuildTestResultSummaryHandler)
+		testResult.GET("/summary/:artifactId/detail", m.getBuildTestResultDetailsHandler)
+
+		testResult.GET("/list-summary", m.getBuildTestResultListSummaryHandler)
+	}
+}
+
+// postBuildTestResultDataHandler godoc
+// @summary Post test result data
+// @tags test-result
+// @accept multipart/form-data
+// @param buildid path int true "Build ID"
+// @param file formData file true "Test result file"
```

Comment:
Here you only specify the form field `file`, while the `ctxparser.ParseMultipartFormData` reads from all files.

According to https://swagger.io/docs/specification/describing-request-body/file-upload/ to allow multiple files to be uploaded, you specify it as an array and then reuse the same form field name but upload multiple files.

Not sure if Swaggo supports it, but maybe the syntax is like so?

```go
// @param files formData []file true "Test result files"
```

And then I suggest you update ParseMultipartFormData to only parse a given form field's files.

Reply:
Can't come up with any security issues.

As the swaggo issue is only for the documentation and not for the actual behavior, I suggest changing `ParseMultipartFormData` to only read from a given form field name (as I mentioned earlier), but then read all files from that field instead of only the first one in the array as it's doing right now.

Then once swaggo supports OpenAPI v3 we can update the field to expect an array of files instead of a single file.

As for the current usage in other projects: AFAIK it's one file per request.

Reply:
> As the swaggo issue is only for the documentation and not for the actual behavior, I suggest changing `ParseMultipartFormData` to only read from a given form field name (as I mentioned earlier), but then read all files from that field instead of only the first one in the array as it's doing right now.

Alright, will leave the methods as-is then, as there is the possibility of multiple files.

Reply:
Done in 4a9871d